services/QuillLMS/app/models/activity_session.rb

```diff
@@ -158,8 +165,18 @@ def finished?
     state == FINISHED_STATE
   end
 
-  def self.calculate_timespent(time_tracking)
-    time_tracking&.values&.compact&.sum
+  def self.time_tracking_sum(time_tracking)
+    return nil unless time_tracking.respond_to?(:values)
+
+    time_tracking.values.compact.sum
+  end
+
+  def self.calculate_timespent(activity_session, time_tracking)
+    timespent = activity_session&.timespent || time_tracking_sum(time_tracking)
```

Comment:
I'm not sure I understand the disjunction here.  Assuming, the there is cleaned data, why would we prefer `activity_session&.timespent` over `time_tracking_sum(time_tracking)`

Reply:
So this is copying the existing code (and just moving it to the model). I agree, this might be better to change, but I think there may be some reason this was in place (time_tracking is new-ish, so I think it's to avoid overwriting older data). Anyway, I'm aiming to keep this changeset as small as possible, so will skip this potential refactor for now.